### PR TITLE
Left/Right Buttons for IntentLister

### PR DESCRIPTION
**Pull request by williamyao27**
## Pull Request Summary
Adds a Left and Right button to Analysis page that switches between sets of intents to display. Can also use arrow keys on keyboard to activate the buttons. Buttons get disabled when reaching their respective boundary (index is 0 for left, index is within 3 of the end for right).

Several files were changed to achieve this. A notable other change was that BaseButton now accepts a label prop in case there is no text, since it was otherwise impossible for testing to get that button (and also it's good for a11y). But for the left/right buttons, most of the work is done in IntentLister on top of the existing index-related code.

The increment/skip left/right icons are placeholders - we can prob work to find pairs that look more alike.

### TODO:
- [x] Add tests
- [x] Figure out why it doesn't respond to arrow keys, let it accept A/D as well
- [x] Update arrow icons to be thicker
- [x] Simultaneous key press
- [x] Validation for having aria label if no text in BaseButton

### Change types

- [x] Non-breaking change
- [x] Adding tests
- [x] Front end change

### Testing
- [x] I have added testing to this PR 
I don't think I can have testing for the keyboard presses by Thursday, but the increment/decrement methods themselves (as well as the buttons) are tested!

Edit: the skip to front/end have not been tested. Will do testing for that in another PR over the next week.

### How to manually test this PR (QA instructions)
Go to Analysis Page. Use the left/right arrow buttons to switch between intents. Use A/D keys and left/right arrow keys to do the same thing. Hold CTRL while doing either to skip to the extreme in that direction. Notice that the icons change as CTRL is held and released.

## Questions

Please check the code for how key press events are handled! I feel it's ugly but I don't have enough experience to know if/where it can be trimmed down.
I followed a combination of these posts:
https://stackoverflow.com/questions/37440408/how-to-detect-esc-key-press-in-react-and-how-to-handle-it
https://stackoverflow.com/questions/5597060/detecting-arrow-key-presses-in-javascript

Please see this post for how to handle multi-key shortcuts:
https://www.gavsblog.com/blog/detect-single-and-multiple-keypress-events-javascript

Please see this post for conditional prop requirement added to BaseButton:
https://stackoverflow.com/questions/42299335/react-props-set-isrequired-on-a-prop-if-another-prop-is-null-empty

### CODE REVIEW

**Review by emilylwan — APPROVED** [review]
Code looks good to me!

File: front-end/src/components/AnalysisPage.jsx [hunk 1]
```diff
@@ -42,10 +73,13 @@ class AnalysisPage extends Component {
...
                     text="Return to Main Page" 
-                    size="sm" />
+                    size="sm"
+                  />
                 </div>
                 <div className="w-4/5 h-4/5 mx-auto">
-                  <IntentLister intents={fakeIntents} index={0} />
+                  <IntentLister
+                    intents={fakeIntents}
+                  />
```
Comment (WeissSa) [inline front-end/src/components/AnalysisPage.jsx:82]: nit (for your follow up PR): this should be on 1 line

File: front-end/src/components/BaseButton.jsx [hunk 2]
```diff
@@ -51,10 +52,20 @@ BaseButton.propTypes = {
     size: PropTypes.string,
     isDisabled: PropTypes.bool,
     icon: PropTypes.object,
+    label: function(props) {
```
Comment (WeissSa) [inline front-end/src/components/BaseButton.jsx:55]: (another nit for follow up. Assume all nits are follow up PR stuff actually): arrow function please :pleading_face:

File: front-end/src/components/IntentLister.jsx [hunk 3]
```diff
@@ -1,9 +1,98 @@
...
+            keysHeld: {
+                ...this.state.keysHeld,
+                [event.key]: true,
+            }
+        });
+
+        // Map keys to methods
+        switch(event.key) {
+            case "ArrowLeft":
+            case "a":
```
Comment (WeissSa) [inline front-end/src/components/IntentLister.jsx:41]: I am sad that this is the official syntax :(

**Review by helias22 — APPROVED** [review]
Wow this is pretty cool 😎

File: front-end/src/components/IntentLister.jsx [hunk 4]
```diff
@@ -1,9 +1,98 @@
...
+        document.addEventListener("keyup", this.handleKeyUp, false);
+    }
+
+    componentWillUnmount() {
+        document.removeEventListener("keydown", this.handleKeyDown, false);
+        document.removeEventListener("keyup", this.handleKeyUp, false);
+    }
+
+    handleKeyDown(event) {
+        // Track that this key has been held down
```
Comment (WeissSa) [inline front-end/src/components/IntentLister.jsx:29]: nit: add why

File: front-end/src/components/IntentLister.jsx [hunk 5]
```diff
@@ -1,9 +1,98 @@
...
+            case "d":
+                this.incrementIndex();
+                break;
+            default:
+                break;
+        };
+    }
+
+    handleKeyUp(event) {
+        // Track that this key has been released
```
Comment (WeissSa) [inline front-end/src/components/IntentLister.jsx:54]: nit: add why

**Review by WeissSa — APPROVED** [review]
Amazing first draft. I only have nits. IMO u can merge and fix them in a follow-up

**Pull request merged**